### PR TITLE
[Execution] Reload unexecuted blocks to execution queues on startup

### PR DESCRIPTION
**Pull request by zhangchiqing**
Fix #4920

This PR fixes the issue that on startup, all the unexecuted blocks need to be reloaded into execution queue, and fetch collection for them for execution. Otherwise, the execution might be halt.

### CODE REVIEW

File: engine/execution/ingestion/engine.go
Comment (Kay-Zee) [inline engine/execution/ingestion/engine.go:198]: This can't be hardcoded, since we'll eventually run into this number

File: engine/execution/ingestion/engine.go
Comment (Kay-Zee) [inline engine/execution/ingestion/engine.go:220]:
```suggestion
	for height := lastExecutedHeight; height <= futureHeight; height++ {
```
There is no hard requesting an already executed block, and it's better to overlap, as it solves some current issues with the lastExecutedHeight not being saved idempotently with the rest of the results

**Review by Kay-Zee — CHANGES_REQUESTED** [review]

File: engine/execution/ingestion/engine.go
Comment (zhangchiqing) [inline engine/execution/ingestion/engine.go:242]: We used to fix by starting from the last executed block, which is 1 height lower than the first unexecuted block. 

The reason we needed that fix is that the write operations to update the executed height might be interrupted by a restart, which causes the highest executed height to be inaccurate. Inaccurate means, the 

This new approach no longer depend on the highest executed height. Instead, it goes through each finalized block and pending block, and check if the block has actually been executed before loading them the execution queue.

File: module/synchronization/core.go
Comment (zhangchiqing) [inline module/synchronization/core.go:244]: I saw the log printing the prunedHeights was negative, and realized the math here needs to reversed.

File: engine/execution/state/state.go
Comment (zhangchiqing) [inline engine/execution/state/state.go:56]: Reusable function builds on top of abstractions.

File: engine/execution/ingestion/engine.go
Comment (m4ksio) [inline engine/execution/ingestion/engine.go:221]: Can this loop turn into infinite one? I mean, if no block is found (ie. empty database or some other kind of error) - we will never break out of it.
Its unlikely, but still can happen - shall we add some form of protection, even if its just checking we don't go below zero or so?

**Review by m4ksio — COMMENTED** [review]
I strongly think we need some tests for new functionality

File: engine/execution/ingestion/engine.go
Comment (zhangchiqing) [reply engine/execution/ingestion/engine.go:221]: yeah, i thought about this as well. And I don't think it will be infinite loop.

Normally, the root block is always executed already, so if we reach the root block, it will exit the loop

And the function body also requires the block header at that height to exist. So if it went underflow or no block is found, which both are exception cases, then the query to find block header will return error and then exit the loop. 

But I agree it's probably safer to ensure it's never go below zero.

Comment (zhangchiqing) [reply engine/execution/ingestion/engine.go]:
```
engine/execution/ingestion/engine.go:205:8: SA4003: every value of type uint64 is >= 0 (staticcheck)
        for ; lastExecuted >= 0; lastExecuted-- {
              ^
```
Actually, would cause this error

File: engine/execution/ingestion/engine_test.go
Comment (zhangchiqing) [inline engine/execution/ingestion/engine_test.go:636]: test cases added @m4ksio

File: engine/execution/ingestion/engine.go
Comment (ramtinms) [reply engine/execution/ingestion/engine.go:230]: Does this solve the problem of moving finalized height? (finalized block increased while we where running older blocks)

File: engine/execution/state/state.go
Comment (ramtinms) [inline engine/execution/state/state.go:57]: This is the only part I'm worried about, the block might be executed and state commitment might be there but it doesn't provide any guarantees that Result has been stored, which causes this issue: 
 - if all exec nodes crash at the same time (after state commitment storage and before result storage), consensus nodes will miss a result forever and sealing will halt.

Comment (ramtinms) [reply engine/execution/state/state.go:57]: while we are waiting for the other issue (transactional updates) to happen I think we should check the exec results (or any thing that is the last piece we store) instead of state commitments and make sure we can re-execute an already executed block. an alternative is to always start re-executing the last executed block anyway.

File: engine/execution/ingestion/engine.go
Comment (zhangchiqing) [reply engine/execution/ingestion/engine.go:232]: Good point. 

I think the finalized height is moving because the follower engine started processing blocks, which triggers BlockProcessable events during the reloading. 

I made a fix to [lock the queue during the reloading](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR283-R290)

File: engine/execution/state/state.go
Comment (zhangchiqing) [reply engine/execution/state/state.go:57]: Agreed. I implemented [the alternative](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR292-R318)

File: utils/unittest/mocks/execution_state.go
Comment (Kay-Zee) [inline utils/unittest/mocks/execution_state.go:18]: Could be something a bit more descriptive

File: utils/unittest/mocks/execution_state.go
Comment (Kay-Zee) [inline utils/unittest/mocks/execution_state.go:48]: Any reason this isn't just attached to the `ES` struct?

File: utils/unittest/mocks/protocol_state.go
Comment (Kay-Zee) [inline utils/unittest/mocks/protocol_state.go:21]: if you decide to change the `ES` struct, would suggest changing this as well

File: utils/unittest/mocks/protocol_state.go
Comment (Kay-Zee) [inline utils/unittest/mocks/protocol_state.go:15]: A little bit hesitant on having a mocked Protocol State, but it does make the unit tests much more isolated, which is nice.

**Review by Kay-Zee — APPROVED** [review]

File: utils/unittest/mocks/protocol_state.go
Comment (zhangchiqing) [reply utils/unittest/mocks/protocol_state.go]: If we don't implement this mock, we will be using another mock, which more or less have to re-implement a partial behaviors, which can't be reused. 

But now, this mock can be reused and extended. 

Testing engines' unittests that depend on protocol state could just use this mocked version

**Review by ramtinms — APPROVED** [review]

**Review by AlexHentschel — CHANGES_REQUESTED** [review]
Generally, I really like the structure you have implemented Leo. Very clean :clap:

I have some concerns regarding:
* information consistency during the reloading of the block (multiple calls to `e.state.Final()` which might return different values). 
*  handling of spork's root block seems too much tailored to a genesis block with height 0 and parent block `flow.Zero`. Both conditions don't hold for a general root block of a spork.

File: engine/execution/ingestion/engine.go
Comment (AlexHentschel) [inline engine/execution/ingestion/engine.go:231]: ⚠️ 

I am not quite sure to which locking you refer with your statement: 
> I made a fix to lock the queue during the reloading

The link just brings me to a view of all the changes of `engine/execution/ingestion/engine_test.go `. Assuming you meant this:
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L286-L290

I am not quite sure whether this completely avoids the edge case. If the follower is running, it might call into the Finalizer, which will update the protocol state. If this happens, your calls to `e.state.Final()` might return _different_ snapshots. 
During your reloading, you query `e.state.Final()` multiple times: 
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L191
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L245
I don't think there are any guarantees that both calls return the same block (?)  
* Lets assume the finalized block is at height 10 in your call to `finalizedUnexecutedBlocks`
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L265-L273
* then, in your call to `pendingUnexecutedBlocks` it has increased to 11. 
* I think you will miss block 11 in this case, because `Pending()`  only returns the descendants of  block 11 but not block 11 itself.  

#### Suggestion

you could query the Finalized State at the beginning before this call:
https://github.com/onflow/flow-go/blob/b6a9788dcde9462bdb99bc40894a7cd11a8bfaca/engine/execution/ingestion/engine.go#L136
and pass the returned snapshot into your methods. A state snapshot is by construction immutable. Therefore, you would execute all your operations with one consistent finalized block

File: engine/execution/ingestion/engine.go
Comment (zhangchiqing) [reply engine/execution/ingestion/engine.go:230]: Good catch 👍

File: engine/execution/state/state.go
Comment (zhangchiqing) [reply engine/execution/state/state.go:56]: Yes, I thought about that.

If `IsBlockExecuted` was a method of `ReadOnlyExecutionState`, then each implementation would have to provide the implementation. 

However, `IsBlockExecuted` is just a function build on top of `StateCommitmentByBlockID`, so if you have implemented `StateCommitmentByBlockID`, then you should get `IsBlockExecuted` for free.

File: utils/unittest/mocks/execution_state.go
Comment (zhangchiqing) [reply utils/unittest/mocks/execution_state.go:49]: I don't think It's necessary, since only concurrent write will need protection. 

There is no write here, only one read from the `commits`.

Also see [this experiment](https://play.golang.org/p/f-zC50dt19r), where it's able to read the value without a lock

File: engine/execution/ingestion/engine.go
Comment (zhangchiqing) [reply engine/execution/ingestion/engine.go:236]: The way it works is that, on restart the ingestion engine will reload blocks, and calls the [`mempool.Run`](https://github.com/onflow/flow-go/pull/73/files#diff-e5c0dea70cf688717060471583da13cc5269a6f661e955613ed7ef698354cc1aR297). This call will acquire the lock for reading and writing to the execution queues in the `mempool`. That means, any calls to `mempool.Run` will be blocked until the lock is released. 

And since I made the entire reloading in `mempool.Run`, it blocks other processes and make them wait for it to finish. So during the reloading, if finalization happens and calls the `BlockProcessable`, which calls the `mempool.Run`, this call will then be blocked, until the reloading is finished.

File: utils/unittest/mocks/execution_state.go
Comment (AlexHentschel) [reply utils/unittest/mocks/execution_state.go:49]: ⚠️ 
The [Go Memory Model](https://golang.org/ref/mem) states:
> Note that a read r may observe the value written by a write w that happens concurrently with r. Even if this occurs, it does not imply that reads happening after r will observe writes that happened before w. 

As far as I understand statement, you are claiming that a read would be able to observe the latest concurrent read even without a lock. This is not true. You will be observing _some_ writes, but there is no guarantee which ones and when without a mutex.

**Review by AlexHentschel — APPROVED** [review]
Thanks for the revisions Leo 👍

**Pull request merged**
